Update MLIR hash to latest MLIR

math-fehr opened the pull request.

webmiche commented:
![image](https://user-images.githubusercontent.com/55622065/212116317-c3520765-a4c6-4841-b340-362b69f0908b.png)

webmiche commented:
[RFC](https://discourse.llvm.org/t/rfc-introduce-a-new-dense-array-attribute/63279/8), [commit](https://github.com/llvm/llvm-project/commit/508eb41d82ca)

superlopuh reviewed (approved):
🚢

webmiche commented on xdsl/parser.py:
```diff
@@ -724,6 +725,10 @@ def parse_integer_type():
             return int_type
 
+        # Shorthand for DenseArrayBase
+        if (dense := self.parse_optional_mlir_dense_array_base()) is not None:
```
I dont think you need the `is not None` part, right?

superlopuh replied:
`dense` might not be None but be another falsey value, the `is not None` is better. The type checker prefers it also in most cases

math-fehr replied:
So this is something that I know @georgebisbas has different opinion than me.
I think that `is not None` is more explicit, and way less error prone, since we won't break it because of a refactor for instance. Though yes, it is redundant.

Since we already uniformized the code by removing the `is not None` in the other files, I'll remove it here as well to be coherent!

math-fehr replied:
I would argue that we should always have `not None`, since this is more robust in general, and convey the right information.
We want to check that something is not `None`, not that something is null, if that makes sense. Even if `is null` == `is None` in these cases, if we change our API this will break without any warning/errors.

webmiche commented on xdsl/parser.py:
```diff
@@ -929,6 +934,34 @@ def parse_mlir_float_type(self, skip_white_space: bool = True) -> AnyFloat:
+            return None
+
+        self.parse_char("<")
+        element_type = self.parse_attribute()
```
Maybe check that the parsed attribute makes sense? (Is integer or float type)

webmiche commented on xdsl/parser.py:
```diff
@@ -1000,6 +1033,10 @@ def parse_optional_mlir_attribute(self,
             return vector
 
+        # dense array
+        if (dense := self.parse_optional_mlir_dense_array_base()) is not None:
```
Dito.

webmiche reviewed (approved):
LGTM overall!

webmiche commented:
I worked on the tests a bit. Weirdly enough, there seems to have been a cache miss on one of my recent pushes. Not sure why, looking into it...

webmiche commented:
Currently, there is an issue with initialized global memrefs. As the initial_value they expect a DenseArrayBase or a UnitAttr, but if I give it a DenseArrayBase, it wants an elements attribute.

I changed he initial_value to a unit attr for the moment...

superlopuh commented on xdsl/dialects/memref.py:
```diff
@@ -218,10 +218,10 @@ def verify_(self) -> None:
-                                                 DenseArrayBase):
-            raise Exception(
-                "Global expects an initial value with type DenseArrayBase")
+        #if self.initial_value and not isinstance(self.initial_value,
```
Is this PR still WIP?

webmiche replied:
There is only this issue left, which relates to the issue I mentioned above.

math-fehr commented:
Thanks! I'll take a look at it!

math-fehr commented:
Note, I would like to merge this only after the new parser is merged.
Otherwise, we'll just end up with more changes in the parser patch, which will delay it again!

math-fehr commented:
(Converting to draft while I'm rebasing everything, and fixing bugs)

math-fehr commented:
I rebased it for after the parser was merged, and I fixed the `global` issue (btw, it should have been `dense`, not `array` that was expected).
Is the fix good for you @webmiche? If that's the case I'll merge it

webmiche commented on xdsl/dialects/memref.py:
```diff
@@ -232,17 +233,16 @@ class Global(Operation):
             raise Exception(
-                "Global expects an initial value with type DenseIntOrFPElementsAttr"
-            )
+                "Global initial value is expected to be a dense type")
```
```suggestion
                "Global initial value is expected to be a dense type or a UnitAttr.")
```

webmiche commented on xdsl/dialects/builtin.py:
```diff
@@ -639,6 +639,48 @@ def tensor_from_list(data: List[int] | List[float],
 
 
+@irdl_attr_definition
+class DenseArrayBase(ParametrizedAttribute):
```
I guess this could use some verifiers, do the types even work and so on.

webmiche reviewed (approved):
Nice, LGTM. Removing the mlir converter already makes us less dependent on the bindings. We are btw still using them to check whether we execute certain tests or not, but I guess we can change that in a different PR in the future.

Also, the name `with-bindings` for one of the test folders doesn't make much sense anymore, though it still uses the bindings to check whether to execute or not, so I guess that's fine :sweat_smile:

math-fehr commented:
Should be good now!
I renamed `with-bindings` to `with-mlir`, but I'm not quite sure how we can check the existence of `mlir` without going through the bindings

math-fehr commented:
We could also look at it in another PR ;)

webmiche commented:
Yeah, I agree with that. I guess we can check for mlir-opt being in the path. This is pretty much the remaining condition to run the tests.

Let's move this to a different PR.
